ocaml/typing/typedecl.ml

```diff
@@ -192,47 +197,13 @@ let enter_type rec_flag env sdecl (id, uid) =
     (* We set ~legacy_immediate to true because we're looking at a declaration
        that was already allowed to be [@@immediate] *)
     layout_of_attributes_default
-      ~legacy_immediate:true ~reason:(Type_declaration path)
+      ~legacy_immediate:true ~context:(Type_declaration path)
       ~default:(Layout.any ~why:Initial_typedecl_env)
       sdecl.ptype_attributes
   in
   let decl =
-    { type_params =
-        (* CR layouts: At the moment, we're defaulting type parameters in
-           recursive type declarations to layout value.  We could probably allow
-           (Sort 'l) and default to value if it's not determined by use.
-
-           Richard supplies the following example:
-
-           (* setup: *)
-           type t_void [@@void]
-           type ('a : void) void_t
-
-           (* case 1: *)
-           type 'b t = 'b void_t * t2
-           and t2 = t_void void_t
-
-           (* case 2: *)
-           type 'b t = 'b void_t * t2
-           and t2 = t_void t
-
-           Case 1 is accepted and case 2 is rejected, which isn't the end of the
-           world, but could perhaps be improved.
-        *)
-        (* CR layouts v2: Actually, RAE thinks this is just wrong now, because
-           make_params defaults to a sort variable and this defaults to value.
-           I'm worried that the value here will propagate somewhere and then
-           conflict with an inferred e.g. float64 somewhere. *)
-        List.map
-          (fun (param, _) ->
-             let layout =
-               layout_of_attributes_default ~legacy_immediate:false
-                 ~reason:(Type_parameter (path, parameter_name param))
-                 ~default:(Layout.value ~why:Type_argument)
-                 param.ptyp_attributes
-             in
-             Btype.newgenvar layout)
-          sdecl.ptype_params;
+    { type_params = List.map (fun (ctyp, _) -> ctyp.ctyp_type)
+                      (make_params ~generic:true env path sdecl.ptype_params);
```

Comment:
The problem is really that translation of parameters has become more difficult, and so I don't want to repeat the code. I don't really know what under `transl_type_param` I could strip out.

Note that the new version doesn't actually add these variables to the env, because `generic` is set and adding to the env is skipped in the `generic` case. I've added a comment in `Typetexp.trasnl_type_param_var` explaining this.

Reply:
Discussed in person and found an easier way to do this without `make_params`. Commit coming shortly.